### PR TITLE
Add core::slice::Split::as_slice

### PR DESCRIPTION
**Pull request by JulianKnodt**
This adds a remainder function to the Slice iterator, so that a caller can access unused
elements if iteration stops.

Addresses #91733

### CODE REVIEW

File: library/core/src/slice/iter.rs
Comment (the8472) [inline library/core/src/slice/iter.rs:360]: I think this should be called `as_slice()` instead, since that's the way it's done for `vec::IntoIter` or `slice::Iter`. And then we could also implement `AsRef<[T]>` like they do.

File: library/core/src/slice/iter.rs
Comment (the8472) [inline library/core/src/slice/iter.rs:358]: The comment needs to be expanded. The tests below should be moved to doctests so users can see how it can be used.

**Review by the8472 — CHANGES_REQUESTED** [review]

File: library/core/src/slice/iter.rs
Comment (JulianKnodt) [reply library/core/src/slice/iter.rs:360]: I think this would have to be insta-stable, but if that's fine

Comment (Patrick-Poitras) [reply library/core/src/slice/iter.rs]: It also matches `str::Split`'s version of remainder. (`as_str()`)

Comment (the8472) [reply library/core/src/slice/iter.rs]: > I think this would have to be insta-stable, but if that's fine

New features require a tracking issue. Stabilizing something requires an FCP on a tracking issue so it'll be run by the libs team anyway. To me personally it looks fine and consistent with other Iterators.

Comment (Patrick-Poitras) [reply library/core/src/slice/iter.rs]: I agree, this needs to be run past the libs team, especially since there is maybe an opportunity to unify with #77998.

Comment (JulianKnodt) [reply library/core/src/slice/iter.rs]: Well the issue is that trait implementations are insta-stable, or at least I can't add an unstable implementation. I can make a tracking issue if this would be agreeable to the libs team but I'm not gonna get ahead of myself and do it before approval

**Comment by camelid** [comment]
triage: What's the status of this PR?

**Comment by JohnCSimon** [comment]
triage:
@kennytm this is still waiting on review fyi

File: library/core/src/slice/iter.rs
Comment (yaahc) [reply library/core/src/slice/iter.rs]: I don't see an urgent need to add an `AsRef` impl so I'd prefer to avoid doing that without strong motivation, but renaming the method from `remainder` to `as_slice` for consistency with `as_str()` on `str::Split` sounds like a good idea to me.

**Comment by yaahc** [comment]
Awesome, looks great. Thank you @JulianKnodt ^_^

@bors r+

**Comment by JulianKnodt** [comment]
@klensy 
From a purely performance perspective, it doesn't make sense to wrap an extra layer of `Option` around a slice since `None` here conveys the same thing as empty slice, the only possibility of difference would be if the iterator will return `None` but not been advanced, or if it has completed and returned `None` already.

If you convert it to `None`, you'd still have to check the tag of the enum, so there's no perf gain.

From a reader's point of view, I'm not sure it's semantically more clear to return an optional slice. From seeing it, I'd be surprised that it returned an optional slice, and not be sure when it would return `None`.

**Comment by pnkfelix** [comment]
Visiting for weekly performance triage. 

The perf run says this regressed externs (a secondary benchmark) incr-full in debug and opt profiles by 1.2%

From my reading of a local cachegrind run on the benchmark, the entirety of the regression is from a closure within `rustc_metadata::rmeta::decoder::cstore_impl::provide`.

I suspect that may be consistent with overhead you get from slight changes in where definitions end up in the encoded metadata from a change like this to libstd.

But also, if you look at the graph, it looks like this is variance that immediately resolved itself in the history of that benchmark:

image

@rustbot label: +perf-regression-triaged

**Pull request merged**
